runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
-      elementTag =
+      this.elementTag =
           StateTags.value("element", WindowedValue.getFullCoder(elementCoder, this.windowCoder));
-      DoFnInvoker<InputT, OutputT> invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
-      restrictionTag = StateTags.value("restriction", restrictionCoder);
+      this.restrictionTag = StateTags.value("restriction", restrictionCoder);
     }
 
-    @Override
+    public void setStateInternalsFactory(StateInternalsFactory<String> stateInternalsFactory) {
+      this.stateInternalsFactory = stateInternalsFactory;
+    }
+
+    public void setTimerInternalsFactory(TimerInternalsFactory<String> timerInternalsFactory) {
+      this.timerInternalsFactory = timerInternalsFactory;
+    }
+
+    public void setOutputManager(DoFnRunners.OutputManager outputManager) {
+      this.outputManager = outputManager;
+    }
+
+    public void setMainOutputTag(TupleTag<OutputT> mainOutputTag) {
+      this.mainOutputTag = mainOutputTag;
+    }
+
+    @Setup
     public void setup() throws Exception {
       invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
     }
 
-    @Override
+    @ProcessElement
     public void processElement(final ProcessContext c) {
+      StateInternals<String> stateInternals =
+          stateInternalsFactory.stateInternalsForKey(c.element().key());
+      TimerInternals timerInternals = timerInternalsFactory.timerInternalsForKey(c.element().key());
+
       // Initialize state (element and restriction) depending on whether this is the seed call.
       // The seed call is the first call for this element, which actually has the element.
       // Subsequent calls are timer firings and the element has to be retrieved from the state.
       TimerInternals.TimerData timer = Iterables.getOnlyElement(c.element().timersIterable(), null);
       boolean isSeedCall = (timer == null);
       StateNamespace stateNamespace = isSeedCall ? StateNamespaces.global() : timer.getNamespace();
       ValueState<WindowedValue<InputT>> elementState =
-          c.windowingInternals().stateInternals().state(stateNamespace, elementTag);
+          stateInternals.state(stateNamespace, elementTag);
       ValueState<RestrictionT> restrictionState =
-          c.windowingInternals().stateInternals().state(stateNamespace, restrictionTag);
+          stateInternals.state(stateNamespace, restrictionTag);
       WatermarkHoldState<GlobalWindow> holdState =
-          c.windowingInternals().stateInternals().state(stateNamespace, watermarkHoldTag);
+          stateInternals.state(stateNamespace, watermarkHoldTag);
 
       ElementAndRestriction<WindowedValue<InputT>, RestrictionT> elementAndRestriction;
       if (isSeedCall) {
         // The element and restriction are available in c.element().
         WindowedValue<ElementAndRestriction<InputT, RestrictionT>> windowedValue =
-            Iterables.getOnlyElement(c.element().elementsIterable());
+            implodeWindows(c.element().elementsIterable());
```

Comment:
In addition to explain that you are allowed to do this, please explain why you do it (to protect against future edits).